File: src/pages/api/zscore.ts
// Section F5: Z-Score Indicator
// Norway and US: daily data — z-score based on ~252 observations per year (robust).
// Germany: monthly data — z-score based on ~12 observations per year (flagged in response).

import type { NextApiRequest, NextApiResponse } from "next";
import { fredSeries, nDaysAgo } from "@/lib/fred";
import { nbFetchYield } from "@/lib/norgesbank";
import { getCache, setCache } from "@/lib/cache";
import { zscore, mean, stddev } from "@/lib/stats";

const CACHE_KEY = "zscore";

export interface ZScoreEntry {
  country: string;
  currentYield: number | null;
  zScore: number | null;
  interpretation: "Historically high" | "Historically low" | "Near average" | "Slightly elevated" | "Slightly depressed" | "Unavailable";
  level: "extreme-high" | "elevated" | "normal" | "depressed" | "extreme-low" | "unavailable";
  sparkline: { date: string; z: number }[];
  dataQuality: "daily" | "monthly";
  dataPoints: number;
}

export interface ZScoreResponse {
  entries: ZScoreEntry[];
  updatedAt: string;
}

function interpret(z: number | null): ZScoreEntry["interpretation"] {
  if (z === null) return "Unavailable";
  if (z >= 2)  return "Historically high";
  if (z <= -2) return "Historically low";
  if (z >= 1)  return "Slightly elevated";
  if (z <= -1) return "Slightly depressed";
  return "Near average";
}

function levelOf(z: number | null): ZScoreEntry["level"] {
  if (z === null) return "unavailable";
  if (z >= 2)  return "extreme-high";
  if (z >= 1)  return "elevated";
  if (z <= -2) return "extreme-low";
  if (z <= -1) return "depressed";
  return "normal";
}

function buildEntry(
  country: string,
  vals: { date: string; value: number }[],
  quality: "daily" | "monthly"
): ZScoreEntry {
  if (!vals.length) {
    return { country, currentYield: null, zScore: null, interpretation: "Unavailable",
      level: "unavailable", sparkline: [], dataQuality: quality, dataPoints: 0 };
  }
  const values = vals.map((v) => v.value);
  const m = mean(values);
  const s = stddev(values);
  const current = vals[vals.length - 1].value;
  const z = parseFloat(zscore(current, values).toFixed(2));

  const cutoff90d = nDaysAgo(90);
  const sparkline = vals
    .filter((p) => p.date >= cutoff90d)
    .map((p) => ({ date: p.date, z: parseFloat(((p.value - m) / (s || 1)).toFixed(3)) }));

  return {
    country,
    currentYield: parseFloat(current.toFixed(2)),
    zScore: z,
    interpretation: interpret(z),
    level: levelOf(z),
    sparkline,
    dataQuality: quality,
    dataPoints: vals.length,
  };
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<ZScoreResponse | { error: string }>
) {
  const cached = getCache<ZScoreResponse>(CACHE_KEY);
  if (cached) return res.status(200).json(cached);

  const start1y = nDaysAgo(365 + 10);
  const startDE = nDaysAgo(365 * 5); // monthly series — need many years for ~60 monthly points

  try {
    const [usRaw, deRaw, noRaw] = await Promise.all([
      fredSeries("DGS10", start1y).catch(() => []),
      fredSeries("IRLTLT01DEM156N", startDE).catch(() => []),
      nbFetchYield("10Y", start1y).catch(() => []),
    ]);

    const usVals = usRaw.map((p) => ({ date: p.date, value: parseFloat(p.value) }));
    const deVals = deRaw.map((p) => ({ date: p.date, value: parseFloat(p.value) }));
    const noVals = noRaw.map((p) => ({ date: p.date, value: p.value }));

    const entries: ZScoreEntry[] = [
      buildEntry("Norway",        noVals, "daily"),
      buildEntry("Germany",       deVals, "monthly"),
      buildEntry("United States", usVals, "daily"),
    ];

    const result: ZScoreResponse = { entries, updatedAt: new Date().toISOString() };
    setCache(CACHE_KEY, result);
    res.status(200).json(result);
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
}

File: backend/routers/zscore.py
from datetime import datetime
from typing import Optional

import asyncio
from fastapi import APIRouter
from fastapi.responses import JSONResponse

from lib.cache import get_cache, set_cache
from lib.fred import fred_series, n_days_ago
from lib.norgesbank import nb_fetch_yield
from lib.stats import zscore, mean, stddev

router = APIRouter()
CACHE_KEY = "zscore"


def _interpret(z: Optional[float]) -> str:
    if z is None:
        return "Unavailable"
    if z >= 2:
        return "Historically high"
    if z <= -2:
        return "Historically low"
    if z >= 1:
        return "Slightly elevated"
    if z <= -1:
        return "Slightly depressed"
    return "Near average"


def _level(z: Optional[float]) -> str:
    if z is None:
        return "unavailable"
    if z >= 2:
        return "extreme-high"
    if z >= 1:
        return "elevated"
    if z <= -2:
        return "extreme-low"
    if z <= -1:
        return "depressed"
    return "normal"


def _build_entry(country: str, vals: list[dict], quality: str) -> dict:
    if not vals:
        return {
            "country": country, "currentYield": None, "zScore": None,
            "interpretation": "Unavailable", "level": "unavailable",
            "sparkline": [], "dataQuality": quality, "dataPoints": 0,
        }
    values = [p["value"] for p in vals]
    m = mean(values)
    s = stddev(values)
    current = vals[-1]["value"]
    z = round(zscore(current, values), 2)

    cutoff_90d = n_days_ago(90)
    sparkline = [
        {"date": p["date"], "z": round((p["value"] - m) / (s or 1), 3)}
        for p in vals
        if p["date"] >= cutoff_90d
    ]

    return {
        "country": country,
        "currentYield": round(current, 2),
        "zScore": z,
        "interpretation": _interpret(z),
        "level": _level(z),
        "sparkline": sparkline,
        "dataQuality": quality,
        "dataPoints": len(vals),
    }


@router.get("/zscore")
async def zscore_route():
    cached = get_cache(CACHE_KEY)
    if cached:
        return cached

    start_1y = n_days_ago(365 + 10)
    start_de = n_days_ago(365 * 5)  # monthly series — need many years for robust z-score

    try:
        us_raw, de_raw, no_raw = await asyncio.gather(
            fred_series("DGS10", start_1y),
            fred_series("IRLTLT01DEM156N", start_de),
            nb_fetch_yield("10Y", start_1y),
            return_exceptions=True,
        )
        if isinstance(us_raw, Exception): us_raw = []
        if isinstance(de_raw, Exception): de_raw = []
        if isinstance(no_raw, Exception): no_raw = []

        entries = [
            _build_entry("Norway",        no_raw, "daily"),
            _build_entry("Germany",       de_raw, "monthly"),
            _build_entry("United States", us_raw, "daily"),
        ]

        result = {"entries": entries, "updatedAt": datetime.utcnow().isoformat() + "Z"}
        set_cache(CACHE_KEY, result)
        return result
    except Exception as e:
        return JSONResponse(status_code=500, content={"error": str(e)})
